### ISSUE TITLE
YAML parsing should respect ignored files and folders

### ISSUE THREAD

**Issue by SugoiDev** [JetBrains/resharper-unity#914]
Currently, it seems the new yaml parsing feature will not consider files/folders set in the "Elements to skip" in Editor -> Inspection Settings. 

This is undesirable when using third party assets or "pure data" assets in general.
In my particular case, I have many thousands of prefabs that contain only data (many from third parties) that should never be parsed and I have DotSettings files to ignore all third party assets in general.

Rider should not parse them if they are defined to be skipped, since this takes considerable time and will increase the time it takes to perform other operations.

Note: I'm currently on the latest nightly.

**Comment by citizenmatt** [comment]
The "Elements to skip" setting is only about code analysis - it lists the items that should be skipped during code inspections. This doesn't affect any YAML files, because we not running any inspections on them. You can read more in the last paragraph of [this page](https://www.jetbrains.com/help/resharper/Reference__Options__Code_Inspection__Settings.html).

But we still index all files, even if they're marked "to skip". This is so that refactorings (e.g. renaming a symbol) still work, and don't produce broken code.

The big problem right now is that the YAML support is very naive and unoptimised, Unity projects use many large YAML files, and we're seeing long solution load times, slow find usages and very high memory usage. #921 is trying to fix these problems, and with a bit of luck, this whole issue will be obsolete. There are two main problems that need fixing:

Firstly, memory usage. We're currently building a full syntax tree for the entire contents of every YAML file, and these are getting very big, and using a lot of memory. Instead, we should be building a "sparse" tree that effectively gives you an outline and parses/creates sub-trees on demand. This should greatly reduce the amount of memory being used. #921 has already done some related work to reduce substring allocations, which has a reasonable impact on larger projects.

Secondly, when we look at the contents of the YAML files, we're looking for "references" - nodes in the tree that can resolve to a method or class (e.g. `UnityEvent` target handler methods, or `MonoScript` class references). Right now, we walk the entire syntax tree, and look at every single node, and these are very large files! This is most visible in solution load and find usages. We can be a lot smarter about this, and only look at the sections of the file that possibly contain references we're interested. #921 has already greatly improved solution load time (a test project went from 100 seconds wall clock time to 6ms for that section of solution load!). It needs a similar fix for find usages, and the "sparse" trees will also help here, too.

All of which means that if everything goes according to plan, this issue will be obsolete. Indexing of YAML files will be as quick as the other files in the solution, and you won't notice extra time in solution load.

However, Unity projects do have an awful lot of YAML files, and some of them are very very big, larger than the threshold for files we normally process in Rider/ReSharper, so we'll have to re-evaluate once #921 is done. All of which means that I'm not ruling out adding a "skip indexing YAML files" list :)

**Comment by SugoiDev** [comment]
Excellent explanation, thanks!

I'll be keeping an eye on your optimizations and I hope this issue does become get obsolete.
I still think an option to ignore files that the IDE shouldn't parse would still be useful. No matter how optimized yaml parsing gets, it can't beat code that doesn't run at all!

For more data-drive games, "pure data" object count (and size!) can explore very fast. Creatives sometimes go crazy with stuff, and with nested prefabs we'll probably see even more data files around as prefab variations.

Mesh and particles "holders" are fairly common, and parsing them doesn't bring much to the IDE side.

**Comment by citizenmatt** [comment]
#914 is merged and should be available in the next nightly and EAP 7. Please can you test again and see how things are? There will still be some impact to indexing these files (they are much larger than your average source file) but hopefully there should be a dramatic improvement.

**Comment by sindrijo** [comment]
I had to disable the 'parse text based assets feature' in a project I have that contains a test scene with about 3000 GameObjects each with one component on them... (8mb) as it was making the 'find usages of feature' unusable because Rider became unresponsive and required a restart each time I tried to use it.

The project itself is quite small (less than 100 classes), I did notice rider never went over 1.6 gb of memory usage while it was struggling but the CPU usage spiked high.

(Version: 218.3.0.135)

**Comment by citizenmatt** [comment]
Is it possible to share the project? We're happy to sign an NDA and you can send it to us directly rather than posting here.

When looking in yaml files, the size of the project isn't really a factor - it's the size and number (and complexity - prefabs) of the yaml files that are important. How large is the scene, and does it include a lot of prefabs? What is the "8mb" referring to?

**Comment by sindrijo** [comment]
8 mb is referring to the size of the file on disk. I will contact you in an IM.
